beacon-chain/db/kv/finalized_block_roots.go

```diff
@@ -0,0 +1,108 @@
+package kv
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/boltdb/bolt"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	dbpb "github.com/prysmaticlabs/prysm/proto/beacon/db"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
+	"github.com/prysmaticlabs/prysm/shared/traceutil"
+	"go.opencensus.io/trace"
+)
+
+var errMissingParentBlockInDatabase = errors.New("missing block in database")
+
+func updateFinalizedBlockRoots(ctx context.Context, tx *bolt.Tx, checkpoint *ethpb.Checkpoint) error {
+	if !featureconfig.Get().EnableFinalizedBlockRootIndex {
+		return nil
+	}
+
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.updateFinalizedBlockRoots")
+	defer span.End()
+
+	bkt := tx.Bucket(finalizedBlockRootsIndexBucket)
+	blocks := tx.Bucket(blocksBucket)
+
+	root := checkpoint.Root
+	var previousRoot []byte
+	genesisRoot := blocks.Get(genesisBlockRootKey)
+
+	// Walk up the ancestry chain until we reach a block root present in the finalized block roots
+	// index bucket or genesis block root.
+	var walk = func() error {
+		for {
+			if bytes.Equal(root, genesisRoot) {
```

Comment:
this walks all the way back to genesis each time we updated the finalized checkpoint, shouldn't it only walk back to the previous finalized checkpoint.

Reply:
It does, see line 67